src/WireFormat.h

```diff
@@ -1605,6 +1648,182 @@ struct TakeIndexletOwnership {
     } __attribute__((packed));
 };
 
+struct TxParticipant {
+    uint64_t tableId;           // Table Id of the participant object.
+    uint64_t keyHash;           // Key Hash of the participant object.
+    uint64_t rpcId;             // Unique (per transaction) participant id.
+
+    TxParticipant()
+        : tableId()
+        , keyHash()
+        , rpcId()
+    {}
+
+    TxParticipant(uint64_t tableId, uint64_t keyHash, uint64_t rpcId)
+        : tableId(tableId)
+        , keyHash(keyHash)
+        , rpcId(rpcId)
+    {
+    }
+
+    bool operator==(const TxParticipant &other) const {
+        return tableId == other.tableId &&
+               keyHash == other.keyHash &&
+               rpcId == other.rpcId;
+    }
+} __attribute__((packed));
+
+struct TxDecision {
+    static const Opcode opcode = Opcode::TX_DECISION;
+    static const ServiceType service = MASTER_SERVICE;
+
+    enum Decision { COMMIT, ABORT, INVALID };
+
+    struct Request {
+        RequestCommon common;
+        Decision decision;          // Result of a transaction commit attempt.
+        uint64_t leaseId;           // Id of the client lease associated with
+                                    // this transaction.
+        uint32_t participantCount;  // Number of local objects participating TX
+                                    // for this server.
+        // List of local Participants
+    } __attribute__((packed));
+
+    struct Response {
+        ResponseCommon common;
+    } __attribute__((packed));
+};
+
+struct TxPrepare {
+    static const Opcode opcode = Opcode::TX_PREPARE;
+    static const ServiceType service = MASTER_SERVICE;
+
+    /// Type of Tx Operation
+    /// Note: Make sure INVALID is always last.
+    enum OpType { READ, REMOVE, WRITE, INVALID };
+
+    enum Vote { COMMIT, ABORT };
+
+    struct Request {
+        RequestCommon common;
+        ClientLease lease;          // Lease information for the requested
+                                    // transaction.  To ensure prepare requests
+                                    // are linearizable.
+        uint64_t ackId;             // Id of the largest RPC id whose metadata
+                                    // can be garbage-collected.  Used for
+                                    // linearizability.
+        uint32_t participantCount;  // Number of all objects participating TX
+                                    // in whole cluster.
+        uint32_t opCount;
+        // List of all Participants of TX.
+        // List of Ops
+
+        struct ReadOp {
+            OpType type;
+            uint64_t tableId;
+            uint64_t rpcId;
+            uint16_t keyLength;
+            RejectRules rejectRules;
```

Comment:
Let's discuss this: I'm wondering if we should get rid of RejectRules, or replace it with something different.